### ISSUE TITLE
Please use a coherent buildsystem

### ISSUE THREAD

**Issue by Conan-Kudo** [kaldi-asr/kaldi#1582]
The way Kaldi is currently structured makes it very hard to build and package for Linux distributions. I understand you guys don't want to make releases, and that's fine, but would you please consider migrating to a coherent buildsystem that supports system wide installs as well as allowing people to build and install in random places (to be embedded with applications, for instance)?

A couple of suggested options, both of which support Linux, Windows, and Mac:
* [Meson](https://github.com/mesonbuild/meson)
* [CMake](https://cmake.org/)

Either option would make it many times easier for it to be built and included in Linux distributions, other projects, etc. as it's easy to incorporate/submodule or whatnot.

**Comment by danpovey** [comment]
The Kaldi project itself is not the kind of thing you'd ever want to include wholesale into a LInux distribution, as it's a research tool that needs to be constantly updated to be useful.

It's possible that you might want to take *parts* of it, that were dependencies for an application of some kind, and include them.  Like a "kaldi runtime".  That is something we'd like to do at some point.

Anyway, I'm open in principle to switching to CMake at some point if others involved in the project think it's a good idea, but it would be a big undertaking.

**Comment by WilliamTambellini** [comment]
+1 for cmake: I guess many users/developers (like me) are reluctant to try/adopt kaldi because of no standard build system.

**Comment by danpovey** [comment]
@jtrmal is working on using cmake for Windows purposes, but on UNIX we
don't have an immediate plan to use it.  That might change though, if we
like it after getting familiar with it.

Actually I think the desire for this feature largely reflects a
misunderstanding of what Kaldi is and what it is usable for; it's not
something we have been hearing from anyone who actually has experience
using Kaldi.

**Comment by jtrmal** [comment]
I actually have a script that converts the makefiles to cmake and I tested
it under Linux as well. Everything seemed reasonable. However, I have high
priority engagement somewhere else, so I'm not able to pursue it for now.
I actually agree with Dan that the build system is more an issue for users
who do not really understand what kaldi is about.
Y.

**Comment by kkm000** [comment]
I am feeling like I need to write a longer answer as a place for future reference, since this question pops up regularly.

Kaldi use is nominally twofold.

First, you would use it as a toolkit to train an acoustic model (using the word *acoustic* loosely, as you can translate other ML problems on sequences into speech recognition terms). For this task, you need the complete toolkit, with most of its binary tools (500+ of them) and libraries (about 30, when I last checked). You'll need a functional bash and a *nix-style filesystem with working symbolic links. We support building Kaldi out of the box like this, on Linux and OS X. I did some work to make Kaldi work under Cygwin about 2 years ago, and trained a model. There was a lot involved beyond the build process; in fact, I scripted build files for MsBuild, but also had to set up a complex environment, ran everything on Windows elevated (to allow creating NTFS symlinks). Also, you need a decent CUDA-compatible GPU or a few, many CPU cores, and probably 64GB or more of main system RAM. Even if your desktop machine (be it Windows or Linux) is powerful enough, you do not want to tie it up with the model training for days (the biggest of my hapless Windows experiments churned for almost two weeks). And that's the bare minimum; if you are experimenting with hyperparameters and such and want reasonable turn-around time, a computing cluster is the way to go. In any case, we are talking at least some dedicated higher-end hardware, and there is no practical reason to port Kaldi to a non-Linux system for this.

The second type of usage is the decoding, which refers to actually using your model for speech recognition. This is where it makes all the sense to build a library for your system that would use the produced model files to do the work. The computational demand is not at all great; one can squeeze the decoder into the performance bracket of modern ARM CPUs, such as one use by the Pi. But then again, you need a matrix algebra library (we support OpenBLAS, ATLAS and Intel MKL, maybe others), and you want only a subset of Kaldi. Practically speaking, you just take the source files and compile into your project what you need. As @danpovey pointed, this is something we might be looking into packaging in the future. However, realistically speaking, as a developer of the model, you'll probably acquire so much familiarity with Kaldi code and structure that this will be the easiest part. Kaldi is written in the standard C++11 language, without so much of metaprogramming that would throw off a less-than-compliant compiler (speaking primarily of Microsoft's cl here), and, having the training part completed, engineering your own recognizer from the smörgåsbord of Kaldi sources would be an easy part. This is why we do not look at the packaging as a priority task. Someone developing a Kaldi-based recognizer might be packaging and distributing the whole thing, including: the acoustic model files with phonetics and (optionally) speaker adaptation data files; the compiled subset of Kaldi required for run-time which varies based on the AM specifics (SGMM? nnet1, 2, 3?) and the overall specification (e. g., allow end user to generate LMs or only use the packaged ones?); external library code, such as linear algebra, OpenFST, necessary codecs; and whatever else is needed for that particular speech recognizer (again, using the word *speech* loosely, as Kaldi is useful for other seq2seq problems).

**TL;DR**: Kaldi is a tool-building kit for an experienced tool builder, and building and packaging anything that you engineer with it for the target system could potentially vary so wildly that pre-tailoring and pre-packaging its parts would yield only a marginal practical value.

**Comment by danpovey** [comment]
Thanks, @kkm!  I agree with your explanation.

On Thu, Aug 17, 2017 at 4:12 PM, Kirill Katsnelson <notifications@github.com
